boards: arc: add nsim support

vonhust opened the pull request:
This commit adds the support of nsim (arc simulator)

* add nsim runner in west
* add board nsim_em which contains two sub configurations:
   * nsim_em: em with ARC MPUv2
   * nsim_sem: em with SecureShield feature and ARC MPUv3
* add soc snps_nsim
* nsim or nsim lite is required, nsimdrv should be in PATH

Fixes #2975

Signed-off-by: Wayne Ren <wei.ren@synopsys.com>

nashif commented:
@vonhust great stuff, just got it working. Is 'make flash' the right thing though? Can we make this work with 'make run' instead?

nashif commented on boards/arc/nsim_em/nsim_sem.yaml:
```diff
@@ -0,0 +1,10 @@
+identifier: nsim_em
+name: EM Nsim simulator
+type: mcu
```
type should be 'simulation' IMO, then we can use it in sanitycheck like qemu.

ruuddw replied:
To get this in sanitycheck is the main goal, indeed. Some help with that would be appreciated to get all the python script i/o redirection, flushing etc. correct.

nashif replied:
yes, will look into that

nashif commented on boards/arc/nsim_em/doc/board.rst:
```diff
@@ -0,0 +1,79 @@
+   :zephyr-app: samples/synchronization
+   :host-os: unix
+   :board: nsim_em
+   :goals: flash
```
should be 'make run', btw, how do you get out of the simulator? I can't abort it for some reason, what is magic key combo? :-)

ruuddw replied:
The CTRL-] combination redirects input from the simulated UART back to the simulator, and CTRL-C stops the simulation. You might need to type a 'reset' command to get the terminal behave normal again, not sure what magic the UART simulation does.

Another trick: the nsimdrv command invoked in the nsim runner has a timeout parameter "-s 100000000" stops the simulation after 100M cycles.

vonhust replied:
Because,  the virtual uart in nsim will hijack the stdin of console. So "ctrl+c" will be caught by the virtual uart,  not the console(nsimdrv). On the other hand, it seems that the nsim has no direct support of "ctrl+c", i.e., if the nsim is running, no way to pause it by "ctrl+c", no magic key combo found.   

So one way to debug is:  open two consoles, one console for make debugserver, the other console to run xxx-xxx-gdb.

vonhust commented:
@nashif ， it seems that for the runner in west, "make run" is not supportted.  if "west run" is typed, an error msg will come out
`
usage: west [-h] [-z ZEPHYR_BASE] [-v] {flash,debug,debugserver} ...
west: error: argument command: invalid choice: 'run' (choose from 'flash', 'debug', 'debugserver')
`
So, it seems "make run" will not call west, but codes in qemu.cmake.

That's why "make flash" is used for nsim.

nashif commented:
> That's why "make flash" is used for nsim.

ok, will take a look.

vonhust commented:
@nashif  continue for "make run".  "make run" is not intergrated into west tool.  To enable "make run" for nsim,  emu/nsim.cmake should be added. That's complicated compared with add a runner in west.

So my suggestion is to put make run in west tool,  remove files in cmake/emu folder

nashif commented:
> So my suggestion is to put make run in west tool, remove files in cmake/emu folder

right. @mbolivar FYI

nashif commented:
ok, I added a run target with simple patch:

https://gist.github.com/nashif/0943b86ecf61913ef389635fe8ab87ff

nashif reviewed (changes requested):
@vonhust I now have sanitycheck working on top of this PR. Can you please split this commit into multiple commits:
- soc
- board
- driver
- west changes?

Once we have that, I can then submit my changes to support 'make run' and sanitycheck integration.

Thanks

Commit:
arch: arc: add nsim support in soc

add nsim soc support, two configs are contained

* em with arc MPUv2
* em with arc SecureShield and MPUv3

Signed-off-by: Wayne Ren <wei.ren@synopsys.com>

Commit:
boards: arc: add virtual board based on nsim

two configs are contained in this board

* nsim_em: em with arc MPUv2
* nsim_sem: em with arc SecureShield and MPUv3

Signed-off-by: Wayne Ren <wei.ren@synopsys.com>

Commit:
drivers: serial: add virtual uart driver for nsim

In nsim, 'nsim_mem-dev=uart0,base=0xf0000000,irq=24'
is used to enable a virtual uart.

This uart is only used to output now, so only poll output
is implemented.

If required, input and interrupt driven are possible.

Signed-off-by: Wayne Ren <wei.ren@synopsys.com>

Commit:
scripts: west: add nsim runner

add nsim runner in west

Signed-off-by: Wayne Ren <wei.ren@synopsys.com>

vonhust commented:
@nashif  splitt this PR into multiple commits as required

dbkinder commented on boards/arc/nsim_em/doc/board.rst:
```diff
@@ -0,0 +1,79 @@
+Overview
+********
+
+This board configuration will use `Designware ARC nSIM`_ to emulate a virtual ARC EM based board including the following features:
```
Please keep doc line lengths < 80 characters

dbkinder commented on boards/arc/nsim_em/doc/board.rst:
```diff
@@ -0,0 +1,79 @@
+* nsim_em which includes normal em features and ARC MPUv2
+* nsim_sem which includes secure em features and ARC MPUv3
+
+For detailed arc features, please refer :file:`boards/arc/nsim_em/support/nsim_em.props` and :file:`boards/arc/nsim_em/support/nsim_sem.props`.
```
``please refer to``

dbkinder commented on boards/arc/nsim_em/doc/board.rst:
```diff
@@ -0,0 +1,79 @@
+        threadB: Hello World from arc!
+        threadA: Hello World from arc!
+        threadB: Hello World from arc!
+
```
I see comments on the PR asking how to end the simulation, but nothing included in the doc about how to do this.

nashif replied:
the simulator itself provides this information when running, I missed that initially.
